server/src/test/java/org/elasticsearch/common/geo/AbstractTreeTestCase.java

```diff
@@ -0,0 +1,348 @@
+
+        // extent that fully encloses the MultiLine
+        assertRelation(GeoRelation.QUERY_CROSSES, reader, reader.getExtent());
+        if (readerExtent.minX() != Integer.MIN_VALUE && readerExtent.maxX() != Integer.MAX_VALUE
+            && readerExtent.minY() != Integer.MIN_VALUE && readerExtent.maxY() != Integer.MAX_VALUE) {
+            assertRelation(GeoRelation.QUERY_CROSSES, reader, Extent.fromPoints(readerExtent.minX() - 1, readerExtent.minY() - 1,
+                readerExtent.maxX() + 1, readerExtent.maxY() + 1));
+        }
+
+    }
+
+    public void testRandomGeometryIntersection() throws IOException {
+        int testPointCount = randomIntBetween(100, 200);
+        Point[] testPoints = new Point[testPointCount];
+        double extentSize = randomDoubleBetween(1, 10, true);
+        boolean[] intersects = new boolean[testPointCount];
+        for (int i = 0; i < testPoints.length; i++) {
+            testPoints[i] = randomPoint(false);
+        }
+
+        Geometry geometry = randomGeometryTreeGeometry();
+        GeoShapeIndexer indexer = new GeoShapeIndexer(true, "test");
+        Geometry preparedGeometry = indexer.prepareForIndexing(geometry);
+
+        // TODO: support multi-polygons
+        assumeFalse("polygon crosses dateline",
+            ShapeType.POLYGON == geometry.type() && ShapeType.MULTIPOLYGON == preparedGeometry.type());
+
+        for (int i = 0; i < testPointCount; i++) {
+            int cur = i;
+            intersects[cur] = fold(preparedGeometry, false, (g, s) -> s || intersects(g, testPoints[cur], extentSize));
+        }
+
+        for (int i = 0; i < testPointCount; i++) {
+            assertEquals(intersects[i], intersects(preparedGeometry, testPoints[i], extentSize));
+        }
+    }
+
+    private Extent bufferedExtentFromGeoPoint(double x, double y, double extentSize) {
+        int xMin = GeoShapeCoordinateEncoder.INSTANCE.encodeX(Math.max(x - extentSize, -180.0));
+        int xMax = GeoShapeCoordinateEncoder.INSTANCE.encodeX(Math.min(x + extentSize, 180.0));
+        int yMin = GeoShapeCoordinateEncoder.INSTANCE.encodeY(Math.max(y - extentSize, -90));
+        int yMax = GeoShapeCoordinateEncoder.INSTANCE.encodeY(Math.min(y + extentSize, 90));
+        return Extent.fromPoints(xMin, yMin, xMax, yMax);
+    }
+
+    private boolean intersects(Geometry g, Point p, double extentSize) throws IOException {
+        return geometryTreeReader(g, GeoShapeCoordinateEncoder.INSTANCE)
+            .relate(bufferedExtentFromGeoPoint(p.getX(), p.getY(), extentSize)) == GeoRelation.QUERY_CROSSES;
+    }
+
+    private static Geometry randomGeometryTreeGeometry() {
+        @SuppressWarnings("unchecked") Function<Boolean, Geometry> geometry = ESTestCase.randomFrom(
+            GeometryTestUtils::randomLine,
+            GeometryTestUtils::randomPoint,
+            GeometryTestUtils::randomPolygon,
+            GeometryTestUtils::randomMultiLine,
+            GeometryTestUtils::randomMultiPoint
```

Comment:
this is likely updatable after Igor's PR. More tests were added to the GeometryTreeTests as well that may make sense here.